### PR TITLE
Add KafkaReader for processing streaming data with Apache Kafka

### PR DESCRIPTION
**Pull request by yongtang**
This is a proposal to add KafkaReader so that it is possible to read data from Kafka like TextLineReader and TFRecordReader.

Apache Kafka is a widely used distributed streaming platform in open source community. The goal of this fix is to create a contrib Reader ops (inherits ReaderBase and is similiar to TextLineReader / TFRecordReader) so that it is possible to reader Kafka streaming data from TensorFlow in a similiar fashion.

This fix uses a C/C++ Apache Kafka client library librdkafka which is released under the 2-clause BSD license, and is widely used in a number of Kafka bindings such as Go, Python, C#/.Net, etc.

Below is a sample usage:
```python
import tensorflow as tf
import tensorflow.contrib.kafka as kafka

# The name of test:0:0:10
# test: topic
# 0: partition
# 0: offset
# 10: length (or -1 for unlimited)
#
# The parameters for KafkaReader
# servers: bootstrap.servers
# group: group.id
# eof: If True then read will block until the length of the data
#      has been read and ignore EOF error from Kafka

filename_queue = tf.train.string_input_producer(["test:0:0:10"])

reader = kafka.KafkaReader(servers="localhost", group="test", eof=True)
key, value = reader.read(filename_queue)
with tf.Session() as sess:
  coord = tf.train.Coordinator()
  threads = tf.train.start_queue_runners(coord=coord)
  for i in range(10):
    print(sess.run([key, value]))
  coord.request_stop()
  coord.join(threads)
```

NOTE: Because of #1419, the clean_deps of `tensorflow/core:framework` and `tensorflow/core:lib` has been commented out temporarily so that it is possible to build with ReaderBase. Any suggestions to address this issue is welcomed.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by martinwicke** [comment]
@jhseu can you take a look?

File: tensorflow/contrib/kafka/kernels/kafka_reader_ops.cc [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+  Status ReadLocked(string* key, string* value, bool* produced,
+                    bool* at_end) override {
+    if (limit_ >= 0 &&
+        (topic_partition_->offset() >= limit_ || offset_ >= limit_)) {
+      *at_end = true;
+      return Status::OK();
+    }
+    while (true) {
+      std::unique_ptr<RdKafka::Message> message(consumer_->consume(1000));
```
Comment (terrytangyuan) [inline tensorflow/contrib/kafka/kernels/kafka_reader_ops.cc]: Not too familiar with `rdkafkacpp` but any reasons behind this hard-coded 1000? Is this a standard practice?

Comment (yongtang) [reply tensorflow/contrib/kafka/kernels/kafka_reader_ops.cc]: Thanks @terrytangyuan. The `1000` is the timeout (in `ms`) for consume() call to return. The call will repeat in the loop. The value was used as in the demo code of librdkafka.

I have updated the PR so that an optional attribute `timeout` could be passed to control this value (instead of hard coded `1000`). Please take a look.

**Comment by jhseu** [comment]
Seems reasonable to add, but can we implement this as a Dataset instead? See `tensorflow/core/kernels/reader_dataset_ops.cc`

**Comment by jhseu** [comment]
The queue style of reading will be deprecated.

**Comment by yongtang** [comment]
Thanks @jhseu for the feedback! Let me take a look and update the PR accordingly.

**Comment by yongtang** [comment]
It seems that Dataset requires `tensorflow/core/kernels/dataset.cc` to link as a `.so`. However, that is problematic becuase the class of `Dataset` will exist in two places and the kernel will not recognize the one in `.so`.

That is quite similar to the issue faced in  #1419. Not sure how to get around it other than building the KafkaDataset into the core kernel.

**Comment by jhseu** [comment]
@mrry @allenlavoie Derek or Allen, any thoughts on how we might make a dataset in contrib that gets exposed as a shared library? I'm guessing it would just require moving some of the classes over to the framework directory?

**Comment by mrry** [comment]
@jhseu It should be possible in the next week or two. The current layering is bit more intricate than that, but a couple of changes in flight will enable us to remove the current dependency on `core/common_runtime` and allow building `DatasetOpKernel` and `DatasetBase` subclasses in a separate shared library.

**Comment by yongtang** [comment]
Thanks @jhseu @mrry, please keep me posted on the changes in Dataset so that I could update the PR accordingly.

**Comment by drpngx** [comment]
@jhseu @mrry have the changes been pushed? Thanks.

**Comment by mrry** [comment]
Not yet: we're waiting on a CL from @rohan100jain that will sever the dependency between `captured_function.cc` and `tensorflow/core/common_runtime/...`, so we can split all the `tf.contrib.data` kernels into a separate library that only depends on `tensorflow/core/framework/...`.

**Comment by drpngx** [comment]
Thanks! @rohan100jain please ping this thread once ready.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@mrry @rohan100jain was the CL submitted?

**Comment by rohan100jain** [comment]
Yes that change was submitted. https://github.com/tensorflow/tensorflow/commit/9f4118d00fa9eb85f81a4eb3f96a5583ae5afcdc

**Comment by drpngx** [comment]
OK, @jhseu I think it's ready for review.

File: tensorflow/contrib/kafka/kernels/kafka_dataset_ops.cc [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+          topics_(std::move(topics)),
+          servers_(servers),
+          group_(group),
+          eof_(eof),
+          timeout_(timeout) {}
+
+    std::unique_ptr<IteratorBase> MakeIterator(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::TextLine")}));
```
Comment (jhseu) [inline tensorflow/contrib/kafka/kernels/kafka_dataset_ops.cc]: Kafka instead of TextLine?

File: tensorflow/contrib/kafka/BUILD [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        "//tensorflow/python:platform",
+        "//tensorflow/python:state_ops",
+        "//tensorflow/python:training",
+        "//tensorflow/python/data/ops:dataset_ops",
+        "//tensorflow/python/data/ops:iterator_ops",
+        "//tensorflow/python/data/ops:readers",
+    ],
+)
+
+tf_py_test(
```
Comment (jhseu) [inline tensorflow/contrib/kafka/BUILD]: Can you give instructions for running the test in a comment here?

File: tensorflow/contrib/kafka/BUILD [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    srcs = ["kernels/kafka_dataset_ops.cc"],
+    visibility = ["//visibility:public"],
+    deps = [
+        "//tensorflow/core:framework",
+        "//tensorflow/core:lib",
+        "//tensorflow/core:lib_internal",
+        "//tensorflow/core/kernels:bounds_check_lib",
+        "//tensorflow/core/kernels:dataset",
+        "//third_party/eigen3",
+        "@kafka//:kafka",
```
Comment (jhseu) [inline tensorflow/contrib/kafka/BUILD]: Can we make this an optional dependency that's set through ./configure? The code can be wrapped around #ifdef if kafka is enabled. See other directories for examples.

**Comment by yongtang** [comment]
Thanks @jhseu for the review. The PR has been updated with comments addressed. Please take a look and let me know if there are any issues.

**Review by jhseu — APPROVED** [review]

**Comment by jhseu** [comment]
Thanks! This is greatly appreciated.
